src/main/java/bio/terra/service/dataset/DatasetService.java

```diff
@@ -353,6 +355,91 @@ public String addDatasetAssetSpecifications(
         .submit();
   }
 
+  public void validateDatasetAssetSpecification(Dataset dataset, AssetModel assetModel) {
+    // Validate Root Table
+    DatasetTable rootTable =
+        dataset.getTables().stream()
+            .filter(datasetTable -> datasetTable.getName().equals(assetModel.getRootTable()))
+            .findFirst()
+            .orElseThrow(
+                () ->
+                    new InvalidAssetException(
+                        "Root table " + assetModel.getRootTable() + " does not exist in dataset."));
+    // Validate Root Column
+    if (!rootTable.getColumns().stream()
+        .map(c -> c.getName())
+        .toList()
+        .contains(assetModel.getRootColumn())) {
+      throw new InvalidAssetException(
+          "Root column "
+              + assetModel.getRootColumn()
+              + " does not exist in table "
+              + rootTable.getName());
+    }
+    // Validate Tables
+    assetModel
+        .getTables()
+        .forEach(
+            assetTable -> {
+              DatasetTable currentTable =
+                  dataset.getTables().stream()
+                      .filter(datasetTable -> datasetTable.getName().equals(assetTable.getName()))
+                      .findFirst()
+                      .orElseThrow(
+                          () ->
+                              new InvalidAssetException(
+                                  "Table " + assetTable.getName() + " does not exist in dataset."));
+              List<String> datasetTableColumnNames =
+                  currentTable.getColumns().stream().map(c -> c.getName()).toList();
+              assetTable
+                  .getColumns()
+                  .forEach(
+                      assetColumn -> {
+                        if (!datasetTableColumnNames.contains(assetColumn)) {
+                          throw new InvalidAssetException(
```

Comment:
It might be a nice improvement if rather than throwing on the first missing column, we find all missing columns first and then throw with all of them.

This echoes a personal opinion I have on validation: I appreciate as a user when a failed validation message can point out several things I did wrong which allows me to correct them all in one fell swoop, rather than failing on the first error found.  Not always possible (some validations only make sense in a chain) or practical though!

Reply:
Great point -- I've updated with this improvement!
Example output from the new test:
![image](https://user-images.githubusercontent.com/13254229/173616058-90f76a00-8988-445c-918f-91aa965443b8.png)